src/compiler/checker.ts

```diff
@@ -24690,6 +24690,19 @@ namespace ts {
                 // If the current type is a union type, remove all constituents that couldn't be instances of
                 // the candidate type. If one or more constituents remain, return a union of those.
                 if (type.flags & TypeFlags.Union) {
+                    if(candidate.flags & TypeFlags.Union){
+                        const assignableType = mapType(type, type => {
+                            if (!!(type.flags & TypeFlags.String) || !!(type.flags & TypeFlags.Number)) {
+                                return filterType(candidate, candidate => isTypeSubtypeOf(candidate, type));

```

Comment:
You probably wanna use `isRelated` and not `isTypeSubtypeOf`, so you use the comparison the guard asks for.